File: p367e/perfect_square.js
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquareLinear = function (num) {
  for (let i = 0, sq = 0; sq <= num; ++i, sq = i * i) {
    if (sq === num) return true;
  }
  return false;
};
var isPerfectSquareMath = function (num) {
  // a perfect square is 1 + 3 + 5 + 7 ...
  let i = 1;
  while (num > 0) {
    num -= i;
    i += 2;
  }
  return num === 0;
};
var isPerfectSquareNewton = function (num) {
  let r = num;
  while (r * r > num) {
    r = ~~((r + num / r) / 2);
  }
  return r * r === num;
};
// TEST
[
  [1, true],
  [2, false],
  [14, false],
  [16, true],
  [255, false],
  [256, true],
].forEach((t) => {
  console.log('Is', t[0], 'perfect square? ->', t[1]);
  console.assert(isPerfectSquareLinear(t[0]) == t[1]);
  console.assert(isPerfectSquareMath(t[0]) == t[1]);
  console.assert(isPerfectSquareNewton(t[0]) == t[1]);
});

File: p367e/perfect_square.py
class Solution:
    def isPerfectSquareLinear(self, num: int) -> bool:
        n, sq = 0, 0
        while sq <= num:
            n += 1
            sq = n * n
            if sq == num:
                return True
        return False

    def isPerfectSquareMath(self, num: int) -> bool:
        i = 1
        while num > 0:
            num -= i
            i += 2
        return num == 0

    def isPerfectSquareNewton(self, num: int) -> bool:
        x = num
        while x * x > num:
            x = (x + num / x) // 2
        return x * x == num


# TESTS
tests = [
    [1, True],
    [2, False],
    [14, False],
    [16, True],
    [255, False],
    [256, True],
]
for t in tests:
    sol = Solution()
    print("Is", t[0], "perfect square? ->", t[1])
    assert t[1] == sol.isPerfectSquareLinear(t[0])
    assert t[1] == sol.isPerfectSquareMath(t[0])
    assert t[1] == sol.isPerfectSquareNewton(t[0])
